rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+will occur only if the parent process is not `tsh` or `tctl`, to prevent
+infinite loops.
+
+An environment variable `TELEPORT_TOOLS_VERSION` will be introduced to enable
+overriding and using a specific version of client tools. This field can be used
+either as an emergency workaround for a known issue or for debugging purposes
+
+##### Errors and warnings
+
+If the user does not enroll in automatic updates and the cluster administrator
+does not self-manage client tools updates as outlined in [Self-managed client
+tools updates](#self-managed-client-tools-updates), a series of warnings and
+errors with increasing urgency will be shown to the user.
+
+If the version of client tools is within the same major version as advertised
+by the cluster, a warning will be shown to urge the user to enroll in automatic
+updates. Warnings will not prevent the user from using client tools that are
+slightly out of date.
+
+```
+$ tsh login --proxy=proxy.example.com
+Warning: Client tools are out of date, update to vX.Y.Z.
+
+Update Teleport to vX.Y.Z from https://goteleport.com/download or your system
+package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+[...]
+```
+
+If the version of client tools is 1 major version below the version advertised
+by the cluster, a warning will be shown that indicates some functionality may
+not work.
+
+```
+$ tsh login --proxy=proxy.example.com
+WARNING: Client tools are 1 major version out of date, update to vX.Y.Z.
+
+Some functionality may not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+```
+
+If the version of client tools is 2 (or more) versions lower than the version
+advertised by the cluster or 1 (or more) version greater than the version
+advertised by the cluster, an error will be shown and will require the user to
+use the `--skip-version-check` flag.
+
+```
+$ tsh login --proxy=proxy.example.com
+ERROR: Client tools are N major versions out of date, update to vX.Y.Z.
+
+Some functionality will not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+Use the "--skip-version-check" flag to bypass this check and attempt to connect
+to this cluster.
+```
+
+#### Self-managed client tools updates
+
+Cluster administrators that want to self-manage client tools updates will be
+able to watch for changes to client tools versions which can then be used to
+trigger other integrations (using MDM software like Jamf) to update the
+installed version of client tools on endpoints.
+
+```
+$ tctl autoupdate watch
```

Comment:
Will this command have to store the state somewhere to understand that the version changed? I wonder if for simplicity we should just let users write their own automation that polls /ping endpoint for tools version.

Reply:
I was thinking it's just built on-top of the watcher. It can just do an initial fetch + print out changes as they come in.

It feels like a nice little tool to have and makes self-managed updates feel like a first class citizen.